rest/diagnostic_api.go

```diff
@@ -0,0 +1,347 @@
+			}
+			if role == nil {
+				continue
+			}
+			if _, skipMapInit := resp.AdminRoleGrants[roleName]; !skipMapInit && roleHist.Source == channels.AdminGrant {
+				resp.AdminRoleGrants[roleName] = make(map[string]map[string]auth.GrantHistory)
+				resp.AdminRoleGrants[roleName][defaultKeyspace] = make(map[string]auth.GrantHistory)
+			} else if _, skipMapInit := resp.DynamicRoleGrants[roleName]; !skipMapInit && roleHist.Source == channels.DynamicGrant {
+				resp.DynamicRoleGrants[roleName] = make(map[string]map[string]auth.GrantHistory)
+				resp.DynamicRoleGrants[roleName][defaultKeyspace] = make(map[string]auth.GrantHistory)
+			}
+			for channel, chanEntry := range role.Channels() {
+				if channel == channels.DocumentStarChannel {
+					continue
+				}
+				roleChanHistory := roleHist
+				if chanEntry.VbSequence.Sequence > roleHist.Entries[len(roleHist.Entries)-1].StartSeq {
+					roleChanHistory.Entries = append(roleChanHistory.Entries, auth.GrantHistorySequencePair{StartSeq: chanEntry.VbSequence.Sequence, EndSeq: roleHist.Entries[len(roleHist.Entries)-1].EndSeq})
+				}
+				if roleHist.Source == channels.DynamicGrant {
+					if entry, ok := resp.DynamicRoleGrants[roleName][defaultKeyspace][channel]; ok {
+						roleChanHistory.Entries = append(entry.Entries, roleChanHistory.Entries...)
+					}
+				} else {
+					if entry, ok := resp.AdminRoleGrants[roleName][defaultKeyspace][channel]; ok {
+						roleChanHistory.Entries = append(entry.Entries, roleChanHistory.Entries...)
+					}
+				}
+				resp.addRoleGrants(roleName, roleHist.Source, defaultKeyspace, channel, roleChanHistory)
+			}
+			// loop over previous role channels
+			for channel, chanHistory := range role.ChannelHistory() {
+				if chanHistory.Entries[len(chanHistory.Entries)-1].StartSeq < roleHist.Entries[len(roleHist.Entries)-1].StartSeq {
+					chanHistory.Entries[len(chanHistory.Entries)-1].StartSeq = roleHist.Entries[len(roleHist.Entries)-1].StartSeq
+				}
+				resp.addRoleGrants(roleName, roleHist.Source, defaultKeyspace, channel, chanHistory)
+			}
+		}
+		bytes, err := base.JSONMarshal(resp)
+		if err != nil {
+			return err
+		}
+		h.writeRawJSON(bytes)
+		return err
+	}
+	// Rebuild roles and channels
+	for scope, collectionConfig := range user.GetCollectionsAccess() {
+		// skip if scope/collection has been removed
+		if _, ok := h.db.CollectionNames[scope]; !ok {
+			continue
+		}
+		for collectionName, _ := range collectionConfig {
+			if _, ok := h.db.CollectionNames[scope][collectionName]; !ok {
+				continue
+			}
+			err = authenticator.RebuildCollectionChannels(userPrinc, scope, collectionName)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if err := authenticator.RebuildRoles(user); err != nil {
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+	for roleName, roleEntry := range user.RoleNames() {
+		role, err := h.db.Authenticator(h.ctx()).GetRole(roleName)
+		if err != nil {
+			return err
+		}
+		if role == nil {
+			continue
+		}
```

Comment:
How would it be possible to find a user that has a role and not have a role exist?

This is probably worth adding a test for:

1. add a role
2. add a user with that role 
3. delete role
4. run _all_channels

At this point, does the role exist in `user.RoleNames()`? If it doesn't, then this circumstance might be worth logging at info for. Could this happen in the case that the role is deleted in flight of this function? If it does, then I'd add a comment here why we aren't logging.